File: ruby-call-signature/src/main/java/org/jetbrains/ruby/codeInsight/types/signature/MethodInfo.java
package org.jetbrains.ruby.codeInsight.types.signature;

import org.jetbrains.annotations.NotNull;

public final class MethodInfo {
    @NotNull
    private final ClassInfo myClassInfo;
    @NotNull
    private final String myName;
    @NotNull
    private final RVisibility myVisibility;

    public MethodInfo(@NotNull ClassInfo classInfo, @NotNull String name, @NotNull RVisibility visibility) {
        myClassInfo = classInfo;
        myName = name;
        myVisibility = visibility;
    }

    @NotNull
    public ClassInfo getClassInfo() {
        return myClassInfo;
    }

    @NotNull
    public String getName() {
        return myName;
    }

    @NotNull
    public RVisibility getVisibility() {
        return myVisibility;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        final MethodInfo that = (MethodInfo) o;

        if (!myClassInfo.equals(that.myClassInfo)) return false;
        if (!myName.equals(that.myName)) return false;
        return myVisibility == that.myVisibility;
    }

    @Override
    public int hashCode() {
        int result = myClassInfo.hashCode();
        result = 31 * result + myName.hashCode();
        result = 31 * result + myVisibility.hashCode();
        return result;
    }

    public enum RVisibility {
        PRIVATE(0, "PRIVATE"),
        PROTECTED(1, "PROTECTED"),
        PUBLIC(2, "PUBLIC");

        @NotNull
        private final String myPresentableName;
        private final byte myValue;

        RVisibility(final int value, @NotNull final String presentableName) {
            myValue = (byte) value;
            myPresentableName = presentableName;
        }

        @NotNull
        public String getPresentableName() {
            return myPresentableName;
        }
    }
}

File: ruby-call-signature/src/main/java/org/jetbrains/ruby/codeInsight/types/signature/MethodInfo.kt
package org.jetbrains.ruby.codeInsight.types.signature

interface MethodInfo {
    val classInfo: ClassInfo
    val name: String
    val visibility: RVisibility
    val location: Location?

    data class Impl(override val classInfo: ClassInfo,
                    override val name: String,
                    override val visibility: RVisibility = RVisibility.PUBLIC,
                    override val location: Location? = null) : MethodInfo

    fun validate(): Boolean {
        if (name.length > LENGTH_OF_NAME) {
            return false
        }
        val loc = location
        if (loc == null || loc.path.length > LENGTH_OF_PATH) {
            return false
        }
        return classInfo.validate()
    }

    companion object {
        val LENGTH_OF_NAME = 100
        val LENGTH_OF_PATH = 1000
    }
}

@JvmOverloads
fun MethodInfo(classInfo: ClassInfo, name: String, visibility: RVisibility, location: Location? = null) =
        MethodInfo.Impl(classInfo, name, visibility, location)

fun MethodInfo(copy: MethodInfo) = with(copy) { MethodInfo.Impl(ClassInfo(classInfo), name, visibility, location) }

data class Location(val path: String, val lineno: Int)

enum class RVisibility constructor(val value: Byte, val presentableName: String) {
    PRIVATE(0, "PRIVATE"),
    PROTECTED(1, "PROTECTED"),
    PUBLIC(2, "PUBLIC");
}